Started hybrid grid points processor

jordansmithgoogle opened the pull request:
The HybridGridPointsProcessor inserts range data into a hybrid grid and serializes as a proto to a file.

SirVer commented on cartographer/io/hybrid_grid_points_processor.cc:
```diff
@@ -0,0 +1,103 @@
+    std::unique_ptr<mapping_3d::RangeDataInserter>
+    range_data_inserter,
+    const string& output_filename,
+    FileWriterFactory file_writer_factory,
```
You use the factory exactly once in Flush. No need to save it in the class, instead pass in a FileWriter in the constructor and only take the FileFactory in the FromDictionary method.

jordansmithgoogle replied:
Done

SirVer commented on cartographer/io/hybrid_grid_points_processor.cc:
```diff
@@ -0,0 +1,103 @@
+      output_filename_(output_filename),
+      file_writer_factory_(file_writer_factory),
+      next_(next) {
+  hybrid_grid_ = common::make_unique<mapping_3d::HybridGrid>(
```
No need for unique_ptr<>. Make the HybridGrid a non-ptr member in your class and create it in the construction list.

jordansmithgoogle replied:
Done

SirVer commented on cartographer/io/hybrid_grid_points_processor.cc:
```diff
@@ -0,0 +1,103 @@
+
+void HybridGridPointsProcessor::Process(
+    std::unique_ptr<PointsBatch> batch) {
+  const std::unordered_set<string> frame_id_set = {
```
Do not filter on the frame_id? And for sure do not hard code these.

I think this should either be done via a commandline option (related  https://github.com/googlecartographer/cartographer_ros/issues/238) or by having a PointsProcessor that selectively drops frame ids. This was already a bit awkward in the ColoringPointsProcessor. What do you think?

jordansmithgoogle replied:
Done

SirVer commented on cartographer/io/hybrid_grid_points_processor.h:
```diff
@@ -0,0 +1,54 @@
+#ifndef CARTOGRAPHER_IO_HYBRID_GRID_POINTS_PROCESSOR_H_
+#define CARTOGRAPHER_IO_HYBRID_GRID_POINTS_PROCESSOR_H_
+
+// Library used for processing points processor pipeline.
```
Remove comment?

jordansmithgoogle replied:
Done

SirVer commented on cartographer/io/hybrid_grid_points_processor.h:
```diff
@@ -0,0 +1,54 @@
+namespace cartographer {
+namespace io {
+
+// Creates a geolocated hybrid grid of the points with voxels being
```
Remove 'geolocated'

jordansmithgoogle replied:
I would like to save a geolocated hybrid grid.  That is a geodetic pose {{lat, lng, alt}, {roll, pitch, yaw}} for the center of the hybrid grid to place it on the surface of the Earth.  The cartographer.mapping_3d.proto.HybridGrid message does not have fields to encode this information.  I proposed a GeolocatedOctree proto that does encode this information.  Perhaps we could wrap the HybridGrid proto with a GeolocatedHybridGrid proto to place it on the surface of the Earth?

SirVer replied:
This is not in this PR, though. And Cartographer always works in a metric space, there is no geodetic pose in the system anywhere, so I do not think this is a good fit for this repository. Jie wanted to do something similar before and side-loaded the information in google3, I suggest the same here. 

Remove 'geolocated', since this is not geolocated.

SirVer commented on cartographer/io/hybrid_grid_points_processor.h:
```diff
@@ -0,0 +1,54 @@
+namespace io {
+
+// Creates a geolocated hybrid grid of the points with voxels being
+// 'voxel_size' big.
```
Also mention that you require `range_data_inserter` options in your dictionary.

jordansmithgoogle replied:
Done

SirVer commented on cartographer/io/hybrid_grid_points_processor.h:
```diff
@@ -0,0 +1,54 @@
+class HybridGridPointsProcessor : public PointsProcessor {
+ public:
+  constexpr static const char* kConfigurationFileActionName =
+      "write_geolocated_hybrid_grid";
```
remove geolocated.

jordansmithgoogle replied:
Done

SirVer commented on cartographer/io/hybrid_grid_points_processor.h:
```diff
@@ -0,0 +1,54 @@
+  FlushResult Flush() override;
+
+ private:
+  std::unique_ptr<mapping_3d::RangeDataInserter> range_data_inserter_;
```
If you hand in the RangeDataInserterOptions instead of the constructed object, you do not require the unique_ptr<> which has slightly better semantics IMHO.

jordansmithgoogle replied:
Done

SirVer commented:
Hey Jordan! Thanks for swinging by :)

I rebased the branch on `master` and fixed some nits while reviewing in 29bc190fb74a962d4c7cf9a55cb386f8ec966f05. I pushed these changes back in your branch.  You will need to `git pull --force` to get my changes (this will overwrite your master branch, make sure you do not loose important changes!). Side note: I usually prefer working in feature branches for this, see [the GitHub flow](https://guides.github.com/introduction/flow/).

I also left 8 in-code comments - GitHub claims they are on an outdated diff, but they still apply. Please consider them.

jordansmithgoogle commented:
Holger,
  How do you run the equivalent of "git5 fix" on a github git repository?
I think that in the future, you should comment on my changes, and I should
make the edits my git client?  Thanks.

Jordan

On Fri, Jun 9, 2017 at 2:04 AM, Holger Rapp <notifications@github.com>
wrote:

> Hey Jordan! Thanks for swinging by :)
>
> I rebased the branch on master and fixed some nits while reviewing in
> 29bc190
> <https://github.com/googlecartographer/cartographer/commit/29bc190fb74a962d4c7cf9a55cb386f8ec966f05>.
> I pushed these changes back in your branch. You will need to git pull
> --force to get my changes (this will overwrite your master branch, make
> sure you do not loose important changes!). Side note: I usually prefer
> working in feature branches for this, see the GitHub flow
> <https://guides.github.com/introduction/flow/>.
>
> I also left 8 in-code comments - GitHub claims they are on an outdated
> diff, but they still apply. Please consider them.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/googlecartographer/cartographer/pull/323#issuecomment-307336781>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbuVPq1XGqjeAzbQnI3LPMr7OXYak3hnks5sCQq2gaJpZM4N0WbI>
> .
>

SirVer commented on cartographer/io/hybrid_grid_points_processor.cc:
```diff
@@ -0,0 +1,85 @@
+    const PointsBatch& points_batch) {
+  auto range_data = common::make_unique<sensor::RangeData>();
+  range_data->origin = points_batch.origin;
+  for (const Eigen::Vector3f& point : points_batch.points) {
```
The function is then two lines (one if you use {} initializer), how about you inline it?

jordansmithgoogle replied:
I inlined using {}

SirVer commented on cartographer/io/hybrid_grid_points_processor.cc:
```diff
@@ -0,0 +1,85 @@
+  std::unique_ptr<sensor::RangeData> range_data_local =
+      RangeDataFromPointsBatch(*batch);
+  range_data_inserter_.Insert(*range_data_local, hybrid_grid_.get());
+
```
remove empty line

jordansmithgoogle replied:
Done

jordansmithgoogle commented on cartographer/io/hybrid_grid_points_processor.h:
```diff
@@ -0,0 +1,51 @@
+namespace io {
+
+// Creates a hybrid grid of the points with voxels being 'voxel_size'
+// big.  'range_data_inserter' options are used to configure the range
```
Done

jordansmithgoogle commented on cartographer/io/hybrid_grid_points_processor.h:
```diff
@@ -0,0 +1,51 @@
+      FileWriterFactory file_writer_factory,
+      common::LuaParameterDictionary* dictionary, PointsProcessor* next);
+
+  ~HybridGridPointsProcessor() override = default;
```
Done

jordansmithgoogle commented on cartographer/io/hybrid_grid_points_processor.h:
```diff
@@ -0,0 +1,51 @@
+      common::LuaParameterDictionary* dictionary, PointsProcessor* next);
+
+  ~HybridGridPointsProcessor() override = default;
+
```
Done

jordansmithgoogle commented on cartographer/io/hybrid_grid_points_processor.h:
```diff
@@ -0,0 +1,51 @@
+
+ private:
+  mapping_3d::RangeDataInserter range_data_inserter_;
+  std::unique_ptr<mapping_3d::HybridGrid> hybrid_grid_;
```
Done

jordansmithgoogle commented on cartographer/io/hybrid_grid_points_processor.h:
```diff
@@ -0,0 +1,51 @@
+  mapping_3d::RangeDataInserter range_data_inserter_;
+  std::unique_ptr<mapping_3d::HybridGrid> hybrid_grid_;
+  std::unique_ptr<FileWriter> file_writer_;
+  PointsProcessor* const next_;
```
Done

jordansmithgoogle commented on cartographer/io/hybrid_grid_points_processor.cc:
```diff
@@ -0,0 +1,85 @@
+
+#include <memory>
+#include <string>
+#include <unordered_set>
```
Done

jordansmithgoogle commented on cartographer/io/hybrid_grid_points_processor.cc:
```diff
@@ -0,0 +1,85 @@
+
+namespace {
+
+std::unique_ptr<sensor::RangeData> RangeDataFromPointsBatch(
```
I inlined using {}

SirVer commented:
Thanks Jordan!
